### PR TITLE
Fix parsing of the `--bind` option for abstract UNIX sockets

### PR DESCRIPTION
**Pull request by cyraxjoe**
❓ **What kind of change does this PR introduce?**
  - [X] 🐞 bug fix
  - [ ] 🐣 feature
  - [ ] 📋 docs update
  - [X] 📋 tests/coverage improvement
  - [ ] 📋 refactoring
  - [ ] 💥 other

📋 **What is the related issue number (starting with `#`)**

It wasn't reported.

❓ **What is the current behavior?** (You can also link to an open issue here)

The `--bind` option wasn't properly parsed for abstract unix socket.
Something like `cheroot --bind @mysocket` would immediately fail with something like:
```python-traceback
$ python -m cheroot 'myapp.wsgi' --bind @MYSOCK
Keyboard Interrupt: shutting down
Traceback (most recent call last):
  File "/nix/store/nzb7x6kij7mk07m88jwv8m1hhyjzjmb1-python3-3.8.0/lib/python3.8/runpy.py", line 192, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/nix/store/nzb7x6kij7mk07m88jwv8m1hhyjzjmb1-python3-3.8.0/lib/python3.8/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/nix/store/ij4qzqw81lnas4f59g05l3amkh928y4l-python3.8-cheroot-8.2.1/lib/python3.8/site-packages/cheroot/__main__.py", line 6, in <module>
    main()
  File "/nix/store/ij4qzqw81lnas4f59g05l3amkh928y4l-python3.8-cheroot-8.2.1/lib/python3.8/site-packages/cheroot/cli.py", line 234, in main
    raw_args._wsgi_app.server(raw_args).safe_start()
  File "/nix/store/ij4qzqw81lnas4f59g05l3amkh928y4l-python3.8-cheroot-8.2.1/lib/python3.8/site-packages/cheroot/server.py", line 1686, in safe_start
    self.start()
  File "/nix/store/ij4qzqw81lnas4f59g05l3amkh928y4l-python3.8-cheroot-8.2.1/lib/python3.8/site-packages/cheroot/server.py", line 1795, in start
    self.prepare()
  File "/nix/store/ij4qzqw81lnas4f59g05l3amkh928y4l-python3.8-cheroot-8.2.1/lib/python3.8/site-packages/cheroot/server.py", line 1728, in prepare
    info = socket.getaddrinfo(
  File "/nix/store/nzb7x6kij7mk07m88jwv8m1hhyjzjmb1-python3-3.8.0/lib/python3.8/socket.py", line 914, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
OSError: [Errno 16] Device or resource busy
```
Because the server would try to connect using TCP/IP at `MYSOCK` port `None`, given the faulty parsing of the CLI.

❓ **What is the new behavior (if this is a feature change)?**

The `--bind` option now works for abstract UNIX sockets.

📋 **Other information**:

📋 **Checklist**:

  - [X] I think the code is well written
  - [ ] I wrote [good commit messages][1]
  - [ ] I have [squashed related commits together][2] after the changes have been approved
  - [X] Unit tests for the changes exist
  - [ ] Integration tests for the changes exist (if applicable)
  - [x] I used the same coding conventions as the rest of the project
  - [x] The new code doesn't generate linter offenses
  - [ ] Documentation reflects the changes
  - [X] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences

[1]: http://chris.beams.io/posts/git-commit/
[2]: https://github.com/todotxt/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit

---
This change is [Reviewable](https://reviewable.io/reviews/cherrypy/cheroot/248)

### CODE REVIEW

File: cheroot/test/test_cli.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
+from cheroot.cli import (
+    Application,
+    parse_wsgi_bind_addr
+)
+
+
+def test_parse_wsgi_bind_location_for_tcpip():
```
Comment (webknjaz) [inline cheroot/test/test_cli.py]: Could you plz use `@pytest.mark.parametrize` here?

File: cheroot/test/test_cli.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    import sys
+
+    class WSGIAppMock:
+        def application(self):
+            pass
+
+        def main(self):
+            pass
+    try:
+        wsgi_app_mock = WSGIAppMock()
```
Comment (webknjaz) [inline cheroot/test/test_cli.py]: Plz make a fixture providing this.

File: cheroot/test/test_cli.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        sys.modules['mypkg.wsgi'] = wsgi_app_mock
```
Comment (webknjaz) [inline cheroot/test/test_cli.py]: I think it's best to rely on the builtin `monkeypatch` fixture provided by pytest.

File: cheroot/cli.py [hunk 4]
```diff
@@ -137,11 +143,7 @@ def parse_wsgi_bind_location(bind_addr_string):
             return TCPSocket(addr, port)
     except ValueError:
         pass
-
```
Comment (webknjaz) [inline cheroot/cli.py]: I'd keep this empty line in place.

**Comment by webknjaz** [comment]
Hey Joel!

The change looks okay, but we need linters to pass and please use pytest's way of organizing tests as I outlined in the comments.

File: cheroot/test/test_cli.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        app = Application.resolve('mypkg.wsgi')
```
Comment (webknjaz) [inline cheroot/test/test_cli.py]: Please use `parametrize` for this as well.

File: cheroot/test/test_cli.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+def test_parse_wsgi_bind_location_for_unix_socket():
+    assert parse_wsgi_bind_addr('/tmp/cheroot.sock') == '/tmp/cheroot.sock'
+
+
+def test_parse_wsgi_bind_addr_for_abstract_unix_socket():
+    assert parse_wsgi_bind_addr('@cheroot') == '\0cheroot'
+
+
+def test_Aplication_resolve():
+    import sys
```
Comment (webknjaz) [inline cheroot/test/test_cli.py]: Imports like this belong to the top of the module. But maybe you could even avoid it with the use of `monkeypatch` fixture.

**Comment by webknjaz** [comment]
P.S. Running `tox -e pre-commit` does some minor bits of autoformatting automatically.

**Comment by cyraxjoe** [comment]
Hey @webknjaz!

Thank you for your detailed review! 

I'll make the changes later in the day. :+1:

**Comment by cyraxjoe** [comment]
:disappointed:  the idea of using ``sys.modules`` as a way to mock the module loading does not work in python 2. ``import_module`` is implemented differently, it uses ``__import__`` without verifying `sys.modules`.

**Comment by cyraxjoe** [comment]
Note to self: mark the PR as `WIP ` to avoid triggering all the verification unnecessarily. :sweat_smile:

File: cheroot/cli.py [hunk 7]
```diff
@@ -49,6 +49,7 @@ def __init__(self, address, port):
         Args:
             address (str): Host name or IP address
             port (int): TCP port number
+
```
Comment (webknjaz) [inline cheroot/cli.py]: Why is this necessary?

Comment (cyraxjoe) [reply cheroot/cli.py]: I went full _follow the linter_, and that missing blank line caused error `D413` from this list: http://www.pydocstyle.org/en/latest/error_codes.html

Comment (webknjaz) [reply cheroot/cli.py]: Interesting. I guess we'll have to switch this to proper sphinx-recognized params anyway. But that's later.

File: cheroot/cli.py [hunk 8]
```diff
@@ -77,12 +78,12 @@ def resolve(cls, full_path):
...
         app = getattr(import_module(mod_path), app_path or 'application')
-
-        with contextlib.suppress(TypeError):
-            if issubclass(app, server.Gateway):
-                return GatewayYo(app)
-
-        return cls(app)
+        # verify that `app` is a class before the issubclass verification,
+        # otherwise a TypeError will be rised.
+        if inspect.isclass(app) and issubclass(app, server.Gateway):
```
Comment (webknjaz) [inline cheroot/cli.py]: EAFP is more natural to Python. If you worry about the Py2 compat, I could cherry-pick this https://github.com/cherrypy/cheroot/pull/243/files#diff-544070cbdc8d4e10b3faf72ba309abc6R18-R31 shim to master.

File: cheroot/test/test_cli.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+from cheroot.cli import (
+    Application,
+    parse_wsgi_bind_addr,
+)
+
+
+@pytest.mark.parametrize(
+    'raw_bind_addr, expected_bind_addr', (
+        ('192.168.1.1:80', ('192.168.1.1', 80)),
+        ('[::1]:8000', ('::1', 8000)),
```
Comment (webknjaz) [inline cheroot/test/test_cli.py]: This reminds me that we should probably also test catch-all addrs like
```suggestion
        ('[::1]:8000', ('::1', 8000)),
        ('[::]:8000', ('::', 8000)),
        ('12345', (12345)),
```

File: cheroot/test/test_cli.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    ),
+)
+def test_parse_wsgi_bind_addr_for_tcpip(raw_bind_addr, expected_bind_addr):
+    """Check the parsing of the --bind option for TCP/IP addresses."""
+    assert parse_wsgi_bind_addr(raw_bind_addr) == expected_bind_addr
```
Comment (webknjaz) [inline cheroot/test/test_cli.py]: I'd merge these all similar tests into `parametrize` of the first one. Just make sure to use `pytest.param(..., ..., id=)` (http://doc.pytest.org/en/latest/example/parametrize.html#different-options-for-test-ids / https://docs.pytest.org/en/latest/reference.html#pytest-param)

File: cheroot/test/test_cli.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+        """Empty custom method (callable) inside the mocked WSGI app.
+
+        It has an empty body because we are expecting to verify that
+        the same method is return no the actual execution of it.
+        """
+
+
+@pytest.mark.parametrize(
+    'wsgi_app_spec, pkg_name, app_method, mocked_app', (
+        ('mypkg.wsgi', 'mypkg.wsgi', 'application', WSGIAppMock()),
```
Comment (webknjaz) [inline cheroot/test/test_cli.py]: `WSGIAppMock()` is evaluated during the test module import time. Better use a fixture.

File: cheroot/test/test_cli.py [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+    """Check the parsing of the --bind option for UNIX Sockets."""
+    assert parse_wsgi_bind_addr('/tmp/cheroot.sock') == '/tmp/cheroot.sock'
+
+
+def test_parse_wsgi_bind_addr_for_abstract_unix_socket():
+    """Check the parsing of the --bind option for Abstract UNIX Sockets."""
+    assert parse_wsgi_bind_addr('@cheroot') == '\0cheroot'
+
+
+class WSGIAppMock:
```
Comment (webknjaz) [inline cheroot/test/test_cli.py]: This is how you make a fixture with pytest:
```suggestion
@pytest.fixture
def wsgi_app():
    """Return a WSGI app stub."""
    class WSGIAppMock:
	    """Mock of a wsgi module."""

	     def application(self):
	        """Empty application method.
	         Default method to be called when no specific callable
	        is defined in the wsgi application identifier.
	         It has an empty body because we are expecting to verify that
	        the same method is return no the actual execution of it.
	        """

	     def main(self):
	        """Empty custom method (callable) inside the mocked WSGI app.
	         It has an empty body because we are expecting to verify that
	        the same method is return no the actual execution of it.
	        """

	return WSGIAppMock()
```

File: cheroot/test/test_cli.py [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+@pytest.mark.parametrize(
+    'wsgi_app_spec, pkg_name, app_method, mocked_app', (
+        ('mypkg.wsgi', 'mypkg.wsgi', 'application', WSGIAppMock()),
+        ('mypkg.wsgi:application', 'mypkg.wsgi', 'application', WSGIAppMock()),
+        ('mypkg.wsgi:main', 'mypkg.wsgi', 'main', WSGIAppMock()),
+    ),
+)
+def test_Aplication_resolve(
+    monkeypatch,
+    wsgi_app_spec, pkg_name, app_method, mocked_app,
```
Comment (webknjaz) [inline cheroot/test/test_cli.py]: To use the fixture from https://github.com/cherrypy/cheroot/pull/248/files#r347003745:
```suggestion
    wsgi_app_spec, pkg_name, app_method, wsgi_app,
```

File: cheroot/test/test_cli.py [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+):
+    """Check the wsgi application name conversion."""
+    if six.PY2:
```
Comment (webknjaz) [inline cheroot/test/test_cli.py]: This whole monkeypatching can be also moved to the fixture so that only testing remains in the test function: https://github.com/cherrypy/cheroot/pull/248/files#r347003745.

File: cheroot/test/test_core.py [hunk 15]
```diff
@@ -282,8 +282,8 @@ def test_content_length_required(test_client):
 def test_large_request(test_client_with_defaults):
     """Test GET query with maliciously large Content-Length."""
     # If the server's max_request_body_size is not set (i.e. is set to 0)
-    # then this will result in an `OverflowError: Python int too large to convert to C ssize_t`
-    # in the server.
+    # then this will result in an `OverflowError: Python int too large to
+    # convert to C `ssize_t` in the server.
```
Comment (webknjaz) [inline cheroot/test/test_core.py]: Wow, I probably missed this in the previous PR. I'll commit it to master directly because it's completely unrelated to this PR.

File: cheroot/cli.py (hunk 8 above)
Comment (cyraxjoe) [reply cheroot/cli.py]: Yeah the original approach was a plain `try/except TypeError`  and ignore with a comment on why are we ignoring that exception, but I end up preferring the LBFL in this case. But it seem that the `suppress` context manager is used a lot in the other PR. If that's going to be included, then yeah I could use that here and be done by leaving this fragment as it was.

File: cheroot/test/test_cli.py (hunk 9 above)
Comment (cyraxjoe) [reply cheroot/test/test_cli.py]: Yeah... I was just avoiding doing the extra work of thinking of future possible uses cases and all the possible values that urllib may support.

File: cheroot/test/test_cli.py (hunk 12 above)
Comment (cyraxjoe) [reply cheroot/test/test_cli.py]: BTW that unpacking syntax doesn't work on Python 2.

File: cheroot/test/test_cli.py (hunk 11 above)
Comment (cyraxjoe) [reply cheroot/test/test_cli.py]: The mock instance is directly related to the other params. I did liked the idea of passing it along the others params to express the direct relationship.

For example a future param `('a.b:foo', 'a.b', 'foo')` will break the test, because the mock doesn't define a `foo` method. This is not obvious via the implicit mock injection of pytest.

But I understand if you have a strong opinion on going over the pytest mock injection route.

File: cheroot/test/test_cli.py (hunk 12 above)
Comment (webknjaz) [reply cheroot/test/test_cli.py]: Wow, I guess I'm too addicted to py3 :)

File: cheroot/test/test_cli.py (hunk 11 above)
Comment (webknjaz) [reply cheroot/test/test_cli.py]: In this case, you'd have to deal with this on the fixture level. This is a pytest way.

File: cheroot/test/test_core.py (hunk 15 above)
Comment (webknjaz) [reply cheroot/test/test_core.py]: This one's in master already.

File: cheroot/cli.py (hunk 8 above)
Comment (webknjaz) [reply cheroot/cli.py]: It's in master already.

File: cheroot/test/test_cli.py [hunk 16]
```diff
@@ -14,68 +14,77 @@
...
     'raw_bind_addr, expected_bind_addr', (
+        # tcp/ip
         ('192.168.1.1:80', ('192.168.1.1', 80)),
+        # ipv6 ips has to be enclosed in brakets when specified in url form
         ('[::1]:8000', ('::1', 8000)),
+        ('localhost:5000', ('localhost', 5000)),
+        # this is a valid input, but foo gets discarted
+        ('foo@bar:5000', ('bar', 5000)),
+        ('foo', ('foo', None)),
+        ('123456789', ('123456789', None)),
```
Comment (webknjaz) [inline cheroot/test/test_cli.py]: How about `:8080`?

Comment (webknjaz) [reply cheroot/test/test_cli.py]: and catch-all addresses like `0.0.0.0`/`::`?

File: cheroot/test/test_cli.py [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+        ('/tmp/cheroot.sock', '/tmp/cheroot.sock'),
+        ('/tmp/some-random-file-name', '/tmp/some-random-file-name'),
+        # abstract sockets
+        ('@cheroot', '\0cheroot'),
+    ),
+)
+def test_parse_wsgi_bind_addr(raw_bind_addr, expected_bind_addr):
+    """Check the parsing of the --bind option.
+
+    Verify some of the supported addresses and the excpected return value.
```
Comment (webknjaz) [inline cheroot/test/test_cli.py]: It'd be nice to also have a negative test with invalid data. But that could be done separately.
